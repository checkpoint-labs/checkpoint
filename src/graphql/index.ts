import { graphqlHTTP } from 'express-graphql';
import {
  GraphQLID,
  GraphQLInt,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLSchema,
  GraphQLString
} from 'graphql';
import { ResolverContext } from './resolvers';

/**
 * Creates an graphql http handler for the query passed a parameters.
 * Returned middleware can be used with express.
 */
export default function get(query: GraphQLObjectType, context: ResolverContext) {
  const schema = new GraphQLSchema({ query });
  return graphqlHTTP({ schema, context, graphiql: {} });
}

/**
 * This objects name and field maps to the values of the _metadata
 * database store
 *
 */
export const MetadataGraphQLObject = new GraphQLObjectType({
  name: '_Metadata',
  description: 'Core metadata values used internally by Checkpoint',
  fields: {
    id: { type: new GraphQLNonNull(GraphQLID), description: 'example: last_indexed_block' },
    value: { type: GraphQLString }
  }
});

/**
 * This objects name and field maps to the values of the _checkpoints
 * database store. And is used to generate entity queries for graphql
 *
 */
export const CheckpointsGraphQLObject = new GraphQLObjectType({
  name: '_Checkpoint',
  description: 'Contract and Block where its event is found.',
  fields: {
    id: {
      type: new GraphQLNonNull(GraphQLID),
      description: 'id computed as last 5 bytes of sha256(contract+block)'
    },
    block_number: {
      type: new GraphQLNonNull(GraphQLInt)
    },
    contract_address: {
      type: new GraphQLNonNull(GraphQLString)
    }
  }
});
