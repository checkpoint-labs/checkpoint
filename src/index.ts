import Checkpoint from './checkpoint';
export { LogLevel } from './utils/logger';
export { AsyncMySqlPool } from './mysql';
export * from './types';

export default Checkpoint;
