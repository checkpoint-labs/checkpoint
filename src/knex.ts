import fs from 'fs';
import knex, { Knex } from 'knex';
import { ConnectionString } from 'connection-string';

export type KnexType =
  | {
      name: 'integer' | 'bigint' | 'boolean' | 'text' | 'json';
    }
  | {
      name: 'decimal';
      options: [number, number];
    }
  | {
      name: 'float';
      options: [number];
    }
  | {
      name: 'string';
      options: [number];
    };

const PROTOCOLS = {
  mysql: 'mysql',
  postgres: 'pg',
  postgresql: 'pg'
};

const EXTRA_OPTIONS = {
  mysql: {
    supportBigNumbers: true,
    bigNumberStrings: true
  }
};

export function getConnectionData(connectionString: string) {
  const connectionConfig = new ConnectionString(connectionString);
  if (!connectionConfig.protocol || !connectionConfig.hosts || !connectionConfig.path) {
    throw new Error('invalid connection string provided');
  }

  const client = PROTOCOLS[connectionConfig.protocol];
  if (!client) {
    throw new Error(`Supplied protocol ${connectionConfig.protocol} is not supported`);
  }

  const sslConfig: { rejectUnauthorized?: boolean; sslmode?: string; ca?: string } = {};
  if (
    connectionConfig.params?.sslaccept === 'strict' ||
    connectionConfig.params?.ssl === 'rejectUnauthorized'
  ) {
    sslConfig.rejectUnauthorized = true;
  }
  if (connectionConfig.params?.sslmode) {
    sslConfig.sslmode = connectionConfig.params.sslmode;
  }

  if (process.env.CA_CERT) {
    sslConfig.ca = process.env.CA_CERT;
  } else if (process.env.CA_CERT_FILE) {
    sslConfig.ca = fs.readFileSync(process.env.CA_CERT_FILE).toString();
  }

  return {
    client,
    connection: {
      database: connectionConfig.path[0],
      user: connectionConfig.user,
      password: connectionConfig.password,
      host: connectionConfig.hosts[0].name,
      port: connectionConfig.hosts[0].port,
      ssl: Object.keys(sslConfig).length > 0 ? sslConfig : undefined,
      ...EXTRA_OPTIONS[client]
    }
  };
}

export function createKnexConfig(connectionString: string): Knex.Config {
  return getConnectionData(connectionString);
}

export function createKnex(config: string | Knex.Config) {
  const parsedConfig = typeof config === 'string' ? createKnexConfig(config) : config;

  return knex(parsedConfig);
}
