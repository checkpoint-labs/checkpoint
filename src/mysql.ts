import mysql, { PoolConfig, Pool as PoolType, QueryOptions } from 'mysql';
import Pool from 'mysql/lib/Pool';
import Connection from 'mysql/lib/Connection';
import bluebird from 'bluebird';
import { ConnectionString } from 'connection-string';

bluebird.promisifyAll([Pool, Connection]);

/**
 * Type definition for the promisified Pool type.
 *
 * This has to be updated manually with new promisified methods
 * that users will like to access.
 */
export interface AsyncMySqlPool extends PoolType {
  queryAsync: (options: string | QueryOptions, values?: any) => Promise<any>;
}

/**
 * Attempts to connect to the database by the connection string. If no connection string
 * argument is provided, it tries to use the `DATABASE_URL` environment variable
 * as connection string.
 *
 * This returns a mysql pool connection object.
 */
export const createMySqlPool = (connection?: string): AsyncMySqlPool => {
  if (!connection && !process.env.DATABASE_URL) {
    throw new Error(
      'a valid connection string or DATABASE_URL environment variable is required to connect to the database'
    );
  }

  const connectionConfig = new ConnectionString((connection || process.env.DATABASE_URL) as string);
  if (!connectionConfig.hosts || !connectionConfig.path) {
    throw new Error('invalid mysql connection string provided');
  }

  const config: PoolConfig = {
    connectionLimit: 1,
    multipleStatements: true,
    database: connectionConfig.path[0],
    user: connectionConfig.user,
    password: connectionConfig.password,
    host: connectionConfig.hosts[0].name,
    port: connectionConfig.hosts[0].port,
    connectTimeout: 30000, // 30 seconds
    charset: 'utf8mb4'
  };

  return mysql.createPool(config) as AsyncMySqlPool;
};
