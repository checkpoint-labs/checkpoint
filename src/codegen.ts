import {
  GraphQLField,
  GraphQLFloat,
  GraphQLID,
  GraphQLInt,
  GraphQLList,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLScalarType,
  GraphQLString,
  GraphQLType,
  isListType
} from 'graphql';
import pluralize from 'pluralize';
import { GqlEntityController } from './graphql/controller';
import { extendSchema } from './utils/graphql';
import { CheckpointConfig } from './types';

type TypeInfo = {
  type: string;
  initialValue: any;
};

type DecimalTypes = NonNullable<CheckpointConfig['decimal_types']>;

const DEFAULT_DECIMAL_TYPES = {
  Decimal: {
    p: 10,
    d: 2
  },
  BigDecimal: {
    p: 20,
    d: 8
  }
};

export const getTypeInfo = (
  type: GraphQLType,
  decimalTypes: DecimalTypes = DEFAULT_DECIMAL_TYPES
): TypeInfo => {
  if (type instanceof GraphQLNonNull) {
    throw new Error('Type must raw type');
  }

  switch (type) {
    case GraphQLInt:
    case GraphQLFloat:
      return { type: 'number', initialValue: 0 };
    case GraphQLString:
    case GraphQLID:
      return { type: 'string', initialValue: '' };
  }

  if (type instanceof GraphQLScalarType) {
    switch (type.name) {
      case 'BigInt':
        return { type: 'bigint', initialValue: 0 };
      case 'Boolean':
        return { type: 'boolean', initialValue: false };
      case 'Text':
        return { type: 'string', initialValue: '' };
    }

    if (type.name in decimalTypes) {
      return { type: 'string', initialValue: '0' };
    }
  }

  if (type instanceof GraphQLObjectType) {
    return { type: 'string', initialValue: '' };
  }

  if (type instanceof GraphQLList) {
    const nonNullNestedType =
      type.ofType instanceof GraphQLNonNull ? type.ofType.ofType : type.ofType;

    return { type: `${getTypeInfo(nonNullNestedType, decimalTypes).type}[]`, initialValue: [] };
  }

  throw new Error('Unknown type');
};

export const getInitialValue = (
  type: GraphQLType,
  decimalTypes: DecimalTypes = DEFAULT_DECIMAL_TYPES
) => {
  if (!(type instanceof GraphQLNonNull)) {
    return null;
  }

  return getTypeInfo(type.ofType, decimalTypes).initialValue;
};

export const getBaseType = (
  type: GraphQLType,
  decimalTypes: DecimalTypes = DEFAULT_DECIMAL_TYPES
) => {
  const nonNullType = type instanceof GraphQLNonNull ? type.ofType : type;

  return getTypeInfo(nonNullType, decimalTypes).type;
};

export const getJSType = (
  field: GraphQLField<any, any>,
  decimalTypes: DecimalTypes = DEFAULT_DECIMAL_TYPES
) => {
  const nonNullType = field.type instanceof GraphQLNonNull ? field.type.ofType : field.type;
  const isNullable = !(field.type instanceof GraphQLNonNull);
  const isList = nonNullType instanceof GraphQLList;
  const baseType = getBaseType(nonNullType, decimalTypes);

  return { isNullable, isList, baseType };
};

export const codegen = (
  schema: string,
  config: CheckpointConfig,
  format: 'typescript' | 'javascript'
) => {
  const decimalTypes = config.decimal_types || DEFAULT_DECIMAL_TYPES;
  const extendedSchema = extendSchema(schema);
  const controller = new GqlEntityController(extendedSchema, config);

  const preamble = `import { Model } from '@snapshot-labs/checkpoint';\n\n`;

  let contents = `${preamble}`;

  controller.schemaObjects.forEach((type, i, arr) => {
    const modelName = type.name;

    contents += `export class ${modelName} extends Model {\n`;
    contents += `  static tableName = '${pluralize(modelName.toLowerCase())}';\n\n`;

    contents += format === 'javascript' ? `  constructor(id) {\n` : `  constructor(id: string) {\n`;
    contents += `    super(${modelName}.tableName);\n\n`;
    controller.getTypeFields(type).forEach(field => {
      const fieldType = field.type instanceof GraphQLNonNull ? field.type.ofType : field.type;
      if (isListType(fieldType) && fieldType.ofType instanceof GraphQLObjectType) {
        return;
      }

      const rawInitialValue = getInitialValue(field.type, decimalTypes);
      const initialValue = field.name === 'id' ? 'id' : JSON.stringify(rawInitialValue);
      contents += `    this.initialSet('${field.name}', ${initialValue});\n`;
    });
    contents += `  }\n\n`;

    contents +=
      format === 'javascript'
        ? `  static async loadEntity(id) {\n`
        : `  static async loadEntity(id: string): Promise<${modelName} | null> {\n`;
    contents += `    const entity = await super.loadEntity(${modelName}.tableName, id);\n`;
    contents += `    if (!entity) return null;\n\n`;
    contents += `    const model = new ${modelName}(id);\n`;
    contents += `    model.setExists();\n\n`;
    contents += `    for (const key in entity) {\n`;
    contents += `      model.set(key, entity[key]);\n`;
    contents += `    }\n\n`;
    contents += `    return model;\n`;
    contents += `  }\n\n`;

    controller.getTypeFields(type).forEach(field => {
      const fieldType = field.type instanceof GraphQLNonNull ? field.type.ofType : field.type;
      if (isListType(fieldType) && fieldType.ofType instanceof GraphQLObjectType) {
        return;
      }

      const { isNullable, isList, baseType } = getJSType(field, decimalTypes);
      const typeAnnotation = isNullable ? `${baseType} | null` : baseType;

      contents +=
        format === 'javascript'
          ? `  get ${field.name}() {\n`
          : `  get ${field.name}(): ${typeAnnotation} {\n`;
      contents += `    return ${
        isList ? `JSON.parse(this.get('${field.name}'))` : `this.get('${field.name}')`
      };\n`;
      contents += `  }\n\n`;

      contents +=
        format === 'javascript'
          ? `  set ${field.name}(value) {\n`
          : `  set ${field.name}(value: ${typeAnnotation}) {\n`;
      contents += `    this.set('${field.name}', ${isList ? `JSON.stringify(value)` : 'value'});\n`;
      contents += `  }\n\n`;
    });

    contents = contents.slice(0, -1);
    contents += i === arr.length - 1 ? '}\n' : '}\n\n';
  });

  return contents;
};
