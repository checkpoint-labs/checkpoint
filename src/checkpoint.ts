import { GetBlockResponse, Provider } from 'starknet';
import { starknetKeccak } from 'starknet/utils/hash';
import { validateAndParseAddress } from 'starknet/utils/address';
import Promise from 'bluebird';
import getGraphQL, { CheckpointsGraphQLObject, MetadataGraphQLObject } from './graphql';
import { GqlEntityController } from './graphql/controller';
import { createLogger, Logger, LogLevel } from './utils/logger';
import { AsyncMySqlPool, createMySqlPool } from './mysql';
import { CheckpointConfig, CheckpointOptions, SupportedNetworkName } from './types';
import { getContractsFromConfig } from './utils/checkpoint';
import { CheckpointRecord, CheckpointsStore, MetadataId } from './stores/checkpoints';
import { GraphQLObjectType } from 'graphql';

export default class Checkpoint {
  public config: CheckpointConfig;
  public writer;
  public schema: string;
  public provider: Provider;

  private readonly entityController: GqlEntityController;
  private readonly log: Logger;
  private readonly sourceContracts: string[];

  private mysqlPool?: AsyncMySqlPool;
  private mysqlConnection?: string;
  private checkpointsStore?: CheckpointsStore;
  private cpBlocksCache: number[] | null;

  constructor(config: CheckpointConfig, writer, schema: string, opts?: CheckpointOptions) {
    this.config = config;
    this.writer = writer;
    this.schema = schema;
    this.entityController = new GqlEntityController(schema);
    this.provider = new Provider({ network: this.config.network as SupportedNetworkName });
    this.sourceContracts = getContractsFromConfig(config);
    this.cpBlocksCache = [];

    this.log = createLogger({
      base: { component: 'checkpoint' },
      level: opts?.logLevel || LogLevel.Error,
      ...(opts?.prettifyLogs
        ? {
            transport: {
              target: 'pino-pretty'
            }
          }
        : {})
    });

    this.mysqlConnection = opts?.dbConnection;
  }

  /**
   * Returns an express handler that exposes a GraphQL API to query entities defined
   * in the schema.
   *
   */
  public get graphql() {
    const entityQueryFields = this.entityController.generateQueryFields();
    const coreQueryFields = this.entityController.generateQueryFields([
      MetadataGraphQLObject,
      CheckpointsGraphQLObject
    ]);

    const querySchema = new GraphQLObjectType({
      name: 'Query',
      fields: {
        ...entityQueryFields,
        ...coreQueryFields
      }
    });

    return getGraphQL(querySchema, {
      log: this.log.child({ component: 'resolver' }),
      mysql: this.mysql
    });
  }

  /**
   * Starts the indexer.
   *
   * The indexer will invoker the respective writer functions when a contract
   * event is found.
   *
   */
  public async start() {
    this.log.debug('starting');

    const blockNum = await this.getStartBlockNum();
    return await this.next(blockNum);
  }

  /**
   * Reset will clear the last synced block informations
   * and force Checkpoint to start indexing from the start
   * block.
   *
   * This will also clear all indexed GraphQL entity records.
   *
   * This should be called when there has been a change to the GraphQL schema
   * or a change to the writer functions logic, so indexing will re-run from
   * the starting block. Also, it should be called the first time Checkpoint
   * is being initialized.
   *
   */
  public async reset() {
    this.log.debug('reset');

    await this.store.createStore();
    await this.store.setMetadata(MetadataId.LastIndexedBlock, 0);

    await this.entityController.createEntityStores(this.mysql);
  }

  /**
   * Registers the blocks where a contracts event can be found.
   * This will be used as a skip list for checkpoints while
   * indexing relevant blocks. Using this seed function can significantly
   * reduce the time for Checkpoint to re-index blocks.
   *
   * This should be called before the start() method is called.
   *
   */
  public async seedCheckpoints(
    checkpointBlocks: { contract: string; blocks: number[] }[]
  ): Promise<void> {
    await this.store.createStore();

    const checkpoints: CheckpointRecord[] = [];

    checkpointBlocks.forEach(cp => {
      cp.blocks.forEach(blockNumber => {
        checkpoints.push({ blockNumber, contractAddress: cp.contract });
      });
    });

    await this.store.insertCheckpoints(checkpoints);
  }

  private async getStartBlockNum() {
    let start = 0;
    let lastBlock = await this.store.getMetadataNumber(MetadataId.LastIndexedBlock);
    lastBlock = lastBlock ?? 0;

    const nextBlock = lastBlock + 1;

    this.config.sources.forEach(source => {
      start = start === 0 || start > source.start ? source.start : start;
    });
    return nextBlock > start ? nextBlock : start;
  }

  private async getNextBlockNumber(blockNumber: number) {}

  private async next(blockNum: number) {
    const checkpointBlock = await this.getNextCheckpointBlock(blockNum);
    if (checkpointBlock) {
      blockNum = checkpointBlock;
    }

    this.log.debug({ blockNumber: blockNum }, 'next block');

    let block: GetBlockResponse;
    try {
      block = await this.provider.getBlock(blockNum);
    } catch (e) {
      this.log.error({ blockNumber: blockNum, err: e }, 'getting block failed... retrying');

      await Promise.delay(12e3);
      return this.next(blockNum);
    }

    await this.handleBlock(block);

    await this.store.setMetadata(MetadataId.LastIndexedBlock, block.block_number);

    return this.next(blockNum + 1);
  }

  private async getNextCheckpointBlock(blockNum: number): Promise<number | null> {
    if (this.cpBlocksCache === null) {
      // cache is null when we can no more find a record in the database
      // so exiting early here to avoid polling the database in subsequent
      // loops.
      return null;
    }

    if (this.cpBlocksCache.length !== 0) {
      return this.cpBlocksCache.shift();
    }

    const checkpointBlocks = await this.store.getNextCheckpointBlocks(
      blockNum,
      this.sourceContracts
    );
    if (checkpointBlocks.length === 0) {
      this.log.info({ blockNumber: blockNum }, 'no more checkpoint blocks in store');
      // disabling cache to stop polling database
      this.cpBlocksCache = null;

      return null;
    }

    this.cpBlocksCache = checkpointBlocks;
    return this.cpBlocksCache.shift();
  }

  private async handleBlock(block: GetBlockResponse) {
    this.log.info({ blockNumber: block.block_number }, 'handling block');

    // @ts-ignore
    for (const receipt of block.transaction_receipts) {
      await this.handleTx(block, block.transactions[receipt.transaction_index], receipt);
    }

    this.log.debug({ blockNumber: block.block_number }, 'handling block done');
  }

  private async handleTx(block, tx, receipt) {
    this.log.debug({ txIndex: tx.transaction_index }, 'handling transaction');

    for (const source of this.config.sources) {
      let foundContractData = false;
      const contract = validateAndParseAddress(source.contract);

      if (contract === validateAndParseAddress(tx.contract_address)) {
        if (tx.type === 'DEPLOY' && source.deploy_fn) {
          foundContractData = true;
          this.log.info(
            { contract: source.contract, txType: tx.type, handlerFn: source.deploy_fn },
            'found deployment transaction'
          );

          await this.writer[source.deploy_fn]({ source, block, tx, receipt, mysql: this.mysql });
        }
      }

      for (const event of receipt.events) {
        if (contract === validateAndParseAddress(event.from_address)) {
          for (const sourceEvent of source.events) {
            if (`0x${starknetKeccak(sourceEvent.name).toString('hex')}` === event.keys[0]) {
              foundContractData = true;
              this.log.info(
                { contract: source.contract, event: sourceEvent.name, handlerFn: sourceEvent.fn },
                'found contract event'
              );

              await this.writer[sourceEvent.fn]({ source, block, tx, receipt, mysql: this.mysql });
            }
          }
        }
      }

      if (foundContractData) {
        await this.store.insertCheckpoints([
          { blockNumber: block.block_number, contractAddress: source.contract }
        ]);
      }
    }

    this.log.debug({ txIndex: tx.transaction_index }, 'handling transaction done');
  }

  private get store(): CheckpointsStore {
    if (this.checkpointsStore) {
      return this.checkpointsStore;
    }

    return (this.checkpointsStore = new CheckpointsStore(this.mysql, this.log));
  }

  private get mysql(): AsyncMySqlPool {
    if (this.mysqlPool) {
      return this.mysqlPool;
    }

    // lazy initialization of mysql connection
    return (this.mysqlPool = createMySqlPool(this.mysqlConnection));
  }
}
