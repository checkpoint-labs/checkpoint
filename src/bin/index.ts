import path from 'path';
import fs from 'fs/promises';
import process from 'process';
import yargs from 'yargs/yargs';
import { hideBin } from 'yargs/helpers';
import { codegen } from '../codegen';
import { CheckpointConfig } from '../types';

const DEFAULT_CONFIG_PATH = 'src/config.json';
const DEFAULT_SCHEMA_PATH = 'src/schema.gql';
const OUTPUT_DIRECTORY = '.checkpoint';

async function generate(schemaFile: string, configFile: string, format: string) {
  if (format !== 'typescript' && format !== 'javascript') {
    throw new Error('Invalid output format');
  }

  console.log('Generating models from schema:', schemaFile);

  const cwd = process.cwd();
  const schemaFilePath = path.join(cwd, schemaFile);
  const configFilePath = path.join(cwd, configFile);

  const schema = await fs.readFile(schemaFilePath, 'utf8');
  const config: CheckpointConfig = await import(configFilePath);

  const generatedModels = codegen(schema, config, format);

  const outputFile = format === 'typescript' ? 'models.ts' : 'models.js';
  const outputPath = path.join(OUTPUT_DIRECTORY, outputFile);

  await fs.mkdir(path.join(cwd, OUTPUT_DIRECTORY), { recursive: true });
  await fs.writeFile(path.join(cwd, outputPath), generatedModels);

  console.log('Models generated to', outputPath);
}

yargs(hideBin(process.argv))
  .command(
    'generate',
    'generate models from schema',
    yargs => {
      return yargs
        .option('schema-file', {
          alias: 's',
          type: 'string',
          default: DEFAULT_SCHEMA_PATH,
          description: 'Schema file path'
        })
        .option('config-file', {
          alias: 'c',
          type: 'string',
          default: DEFAULT_CONFIG_PATH,
          description: 'Config file path'
        })
        .option('output-format', {
          alias: 'f',
          type: 'string',
          default: 'typescript',
          description: 'Output format (typescript or javascript)'
        });
    },
    async argv => {
      try {
        await generate(argv['schema-file'], argv['config-file'], argv['output-format']);
      } catch (err) {
        console.error('Error generating models:', err);
        process.exit(1);
      }
    }
  )
  .demandCommand(1, 'You need to specify a command')
  .parse();
