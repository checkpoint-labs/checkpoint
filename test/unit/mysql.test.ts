import mysql from 'mysql';
import { createMySqlPool } from '../../src/mysql';

jest.mock('mysql');

describe('createMySqlPool()', () => {
  const OLD_ENV = process.env;

  beforeEach(() => {
    jest.resetModules();
    process.env = { ...OLD_ENV };
  });

  afterAll(() => {
    process.env = OLD_ENV;
  });

  it('should throw if no string or environment variable', () => {
    process.env.DATABASE_URL = undefined;

    expect(() => createMySqlPool()).toThrowErrorMatchingSnapshot();
  });

  it('should use environment variables to create connection', () => {
    process.env.DATABASE_URL = 'mysql://root:pass@localhost:3306/env_db';

    createMySqlPool();

    expect(mysql.createPool).toMatchSnapshot();
  });

  it('should use argument string to create connection', () => {
    const connection = 'mysql://root:pass@localhost:3306/string_db';

    createMySqlPool(connection);

    expect(mysql.createPool).toMatchSnapshot();
  });
});
