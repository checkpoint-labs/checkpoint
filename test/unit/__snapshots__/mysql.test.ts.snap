// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createMySqlPool() should throw if no string or environment variable 1`] = `"a valid connection string or DATABASE_URL environment variable is required to connect to the database"`;

exports[`createMySqlPool() should use argument string to create connection 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "charset": "utf8mb4",
        "connectTimeout": 30000,
        "connectionLimit": 1,
        "database": "string_db",
        "host": "localhost",
        "multipleStatements": true,
        "password": "pass",
        "port": 3306,
        "user": "root",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`createMySqlPool() should use environment variables to create connection 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "charset": "utf8mb4",
        "connectTimeout": 30000,
        "connectionLimit": 1,
        "database": "env_db",
        "host": "localhost",
        "multipleStatements": true,
        "password": "pass",
        "port": 3306,
        "user": "root",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
